2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,493 @@
+}
+```
+
+```go
+tx_undelegate(validator_address, delegator_address, amount)
+{
+  unbond(validator_address, delegator_address, amount)
+}
+```
+
+<!-- 
+```go
+tx_redelegate(src_validator_address, dest_validator_address, delegator_address, amount)
+{
+  unbond(src_validator_address, delegator_address, amount)
+  bond(dst_validator_address, delegator_address, amount, unbonding_length)
+}
+```
+-->
+
+```go
+tx_withdraw_unbonds_delegator(delegator_address)
+{
+  forall (validator_address in validators) do
+    withdraw(validator_address, delegator_address)
+}
+```
+
+## PoS functions
+
+
+```go
+//This function is called by transactions tx_self_bond, tx_delegate and tx_redelegate
+//the only possible values for offset_length are pipeline_length and unbonding_length
+func bond(validator_address, delegator_address, amount, offset_length)
+{
+  //QUESTION: What if we remove this and it is up to the user to be smart bonding its tokens
+  if is_validator(validator_address, cur_epoch+offset_length) then
+    //add amount bond to delta at n+offset_length
+    bonds[delegator_address][validator_address].deltas[cur_epoch+offset_length] += amount
+    //debit amount from delegator account and credit it to the PoS account
+    balances[delegator_address] -= amount
+    balances[pos] += amount
+    //compute new total_deltas and write it at n+offset_length
+    var total = read_epoched_field(validators[validator_address].total_deltas, cur_epoch+offset_length, 0)
+    validators[validator_address].total_deltas[cur_epoch+offset_length] = total + amount
+    //update validator's voting_power, total_voting_power and validator_sets at n+offset_length
+    power_before = read_epoched_field(validators[validator_address].voting_power, cur_epoch+offset_length, 0)
+    power_after = update_voting_power(validator_address, cur_epoch+offset_length)
+    update_total_voting_power(cur_epoch+offset_length)
+    update_validator_sets(validator_address, cur_epoch+offset_length, power_before, power_after)
+}
+```
+
+```go
+/* COMMENT
+  two issues:
+  - https://github.com/informalsystems/partnership-heliax/issues/6
+    delbonds and epoch_counter are considered from the unbonding_length offset.
+    This could lead to scenarios in which one starts unbonding before a bond is materialized.
+    Furthermore, there cannot be positive bonds beyond cur_epoch + pipeline_length, so it does not
+    make sense for delbonds.
+    conclusion: intended
+  - https://github.com/informalsystems/partnership-heliax/issues/7
+    there is a problem with unbonding, slashing and total_deltas becoming negative
+    conclusion: it is an actual issue, unresolved
+*/
+//This function is called by transactions tx_unbond, tx_undelegate and tx_redelegate
+func unbond(validator_address, delegator_address, amount)
+{
+  //disallow unbonding if the validator is frozen
+  var state = read_epoched_field(validators[validator_address].state, cur_epoch, ⊥)
+  if (state != frozen) then
```

Comment:
how about jailed?

Reply:
I think this is intended, one should allow delegators to unbond their tokens if the validator gets jalied.